File: problems/tier1-foundation/004-vending-machine/solution.java
// Vending Machine — Solution (Java, State pattern)
import java.util.*;

class VMItem {
    public String name;
    public double price;
    public int quantity;
    public VMItem(String name, double price, int quantity) {
        this.name = name; this.price = price; this.quantity = quantity;
    }
}

class VendingMachine {
    public static final String IDLE = "Idle";
    public static final String PAYMENT_PENDING = "PaymentPending";
    public static final String DISPENSING = "Dispensing";
    public static final String MAINTENANCE = "Maintenance";

    public String state = IDLE;
    public String selectedItem = "";
    public double insertedMoney = 0.0;
    public Map<String, VMItem> inventory = new LinkedHashMap<>();
    public String operatorPin = "1234";

    public VendingMachine() {
        stockDefaults();
    }

    private void stockDefaults() {
        inventory.put("Cola",  new VMItem("Cola",  25.0, 5));
        inventory.put("Chips", new VMItem("Chips", 15.0, 3));
    }

    public void reset() {
        state = IDLE;
        selectedItem = "";
        insertedMoney = 0.0;
        inventory.clear();
        stockDefaults();
    }

    public void selectItem(String item) {
        if (!state.equals(IDLE)) return;
        VMItem inv = inventory.get(item);
        if (inv != null && inv.quantity > 0) {
            selectedItem = item;
            state = PAYMENT_PENDING;
        }
    }

    public void insertMoney(double amount) {
        if (!state.equals(PAYMENT_PENDING)) return;
        insertedMoney += amount;
        if (insertedMoney >= inventory.get(selectedItem).price) {
            state = DISPENSING;
        }
    }

    public void dispense() {
        if (!state.equals(DISPENSING)) return;
        inventory.get(selectedItem).quantity -= 1;
        insertedMoney = 0.0;
        selectedItem = "";
        state = IDLE;
    }

    public void cancel() {
        if (!state.equals(PAYMENT_PENDING)) return;
        insertedMoney = 0.0;
        selectedItem = "";
        state = IDLE;
    }

    public void enterMaintenance(String pin) {
        if (pin.equals(operatorPin) && state.equals(IDLE)) {
            state = MAINTENANCE;
        }
    }

    public void exitMaintenance(String pin) {
        if (pin.equals(operatorPin) && state.equals(MAINTENANCE)) {
            state = IDLE;
        }
    }

    public void restock(String item, int qty) {
        if (!state.equals(MAINTENANCE)) return;
        VMItem inv = inventory.get(item);
        if (inv == null) {
            inv = new VMItem(item, 0.0, 0);
            inventory.put(item, inv);
        }
        inv.quantity += qty;
    }
}

public class Solution {
    private static VendingMachine vm = new VendingMachine();

    public static void reset_service() { vm = new VendingMachine(); }
    public static void reset() { vm.reset(); }
    public static String getState() { return vm.state; }
    public static void selectItem(String item) { vm.selectItem(item); }
    public static void insertMoney(double amount) { vm.insertMoney(amount); }
    public static void dispense() { vm.dispense(); }
    public static void cancel() { vm.cancel(); }
    public static void enterMaintenance(String pin) { vm.enterMaintenance(pin); }
    public static void exitMaintenance(String pin) { vm.exitMaintenance(pin); }
    public static void restock(String item, int qty) { vm.restock(item, qty); }

    public static int vm_get_quantity(String item) {
        VMItem inv = vm.inventory.get(item);
        return inv == null ? -1 : inv.quantity;
    }
}

File: problems/tier1-foundation/004-vending-machine/solution.go
// Vending Machine — State pattern reference solution (Go).
package main

const (
	vmIdle           = "Idle"
	vmPaymentPending = "PaymentPending"
	vmDispensing     = "Dispensing"
	vmMaintenance    = "Maintenance"
)

type vmItem struct {
	name     string
	price    float64
	quantity int
}

type vendingMachine struct {
	state         string
	selectedItem  string
	insertedMoney float64
	inventory     map[string]*vmItem
	operatorPin   string
}

func newVendingMachine() *vendingMachine {
	vm := &vendingMachine{
		state:       vmIdle,
		operatorPin: "1234",
		inventory:   map[string]*vmItem{},
	}
	vm.stockDefaults()
	return vm
}

func (vm *vendingMachine) stockDefaults() {
	vm.inventory["Cola"] = &vmItem{name: "Cola", price: 25.0, quantity: 5}
	vm.inventory["Chips"] = &vmItem{name: "Chips", price: 15.0, quantity: 3}
}

func (vm *vendingMachine) reset() {
	vm.state = vmIdle
	vm.selectedItem = ""
	vm.insertedMoney = 0.0
	vm.inventory = map[string]*vmItem{}
	vm.stockDefaults()
}

func (vm *vendingMachine) selectItem(item string) {
	if vm.state != vmIdle {
		return
	}
	if it, ok := vm.inventory[item]; ok && it.quantity > 0 {
		vm.selectedItem = item
		vm.state = vmPaymentPending
	}
}

func (vm *vendingMachine) insertMoney(amount float64) {
	if vm.state != vmPaymentPending {
		return
	}
	vm.insertedMoney += amount
	if vm.insertedMoney >= vm.inventory[vm.selectedItem].price {
		vm.state = vmDispensing
	}
}

func (vm *vendingMachine) dispense() {
	if vm.state != vmDispensing {
		return
	}
	vm.inventory[vm.selectedItem].quantity--
	vm.insertedMoney = 0.0
	vm.selectedItem = ""
	vm.state = vmIdle
}

func (vm *vendingMachine) cancel() {
	if vm.state != vmPaymentPending {
		return
	}
	vm.insertedMoney = 0.0
	vm.selectedItem = ""
	vm.state = vmIdle
}

func (vm *vendingMachine) enterMaintenance(pin string) {
	if pin == vm.operatorPin && vm.state == vmIdle {
		vm.state = vmMaintenance
	}
}

func (vm *vendingMachine) exitMaintenance(pin string) {
	if pin == vm.operatorPin && vm.state == vmMaintenance {
		vm.state = vmIdle
	}
}

func (vm *vendingMachine) restock(item string, qty int) {
	if vm.state != vmMaintenance {
		return
	}
	if _, ok := vm.inventory[item]; !ok {
		vm.inventory[item] = &vmItem{name: item, price: 0.0, quantity: 0}
	}
	vm.inventory[item].quantity += qty
}

var gVM = newVendingMachine()

func reset_service() {
	gVM = newVendingMachine()
}

func reset() {
	gVM.reset()
}

func getState() string {
	return gVM.state
}

func selectItem(item string) {
	gVM.selectItem(item)
}

func insertMoney(amount float64) {
	gVM.insertMoney(amount)
}

func dispense() {
	gVM.dispense()
}

func cancel() {
	gVM.cancel()
}

func enterMaintenance(pin string) {
	gVM.enterMaintenance(pin)
}

func exitMaintenance(pin string) {
	gVM.exitMaintenance(pin)
}

func restock(item string, qty int) {
	gVM.restock(item, qty)
}

func vm_get_quantity(item string) int {
	if it, ok := gVM.inventory[item]; ok {
		return it.quantity
	}
	return -1
}
